ouroboros-consensus/docs/HardWonWisdom.md

```diff
@@ -0,0 +1,97 @@
+        | NotOrigin for == at ->
+              return
+            $ TPraos.TickedPraosLedgerView
+            $ SL.currentLedgerView shelleyLedgerState
+        | for < maxFor        -> return $ futureLedgerView for
+        | otherwise           -> throwError OutsideForecastRange { ... }
+    where
+      ShelleyLedgerState {shelleyLedgerState} = ledgerState
+
+      globals = shelleyLedgerGlobals cfg
+      swindow = SL.stabilityWindow globals
+      at      = ledgerTipSlot ledgerState
+
+      futureLedgerView :: SlotNo -> Ticked (SL.LedgerView (EraCrypto era))
+      futureLedgerView =
+        either
+          (\e -> error ("futureLedgerView failed: " <> show e))
+          TPraos.TickedPraosLedgerView
+          . SL.futureLedgerView globals shelleyLedgerState
+
+      maxFor :: SlotNo   -- Exclusive upper bound
+      maxFor = addSlots swindow $ succWithOrigin at
+```
+
+When we returned to this code for the first time in a while, we thought it was odd that the code both handles an `Either` in the return type of `SL.futureLedgerView` and also does its own `for < maxFor` check; can't the Ledger code instead return `Left` whenever `for >= maxFor`?
+
+We looked at the upstream code to investigate; how easily could we arrange that?
+The answer is: it could be done, but the core architecture of the State Transition System code in the core Ledger library does not currently naturally allow for that (at least in our subjective opinion of _naturally_ and specifically for the `TICKF` example).
+At the moment, any rule that can _fail_ must also provide a "default" value for the rest of the rule to use (see the `Predicate` constructor of the `Clause` GADT, specifically its argument of type `a`).
+For the Ledger's code implementing the `TICKF` rule to observe the `Either` layer of the `EpochInfo` would require that the rule has some way to continue when the `EpochInfo` query (used to determine if the requested forecast crosses an epoch transition) fails.
+
+It is not obvious how to do that when `EpochInfo` returns `Left`.
+While it may be possible to work something out, such as perhaps the `TICKF` rule emits a new `PredicateFailure` and simply leaves the `NewEpochState` unchanged, no idea seems natural to us.
+They all involve creating an incorrect `LedgerView` and then only throwing it away at the last moment.
+And notably, the rest of the rule could emit additional errors, all of which would be presumably spurious given that we already know the very first check (ie querying the `EpochInfo`) failed.
+
+So instead, for now at least, the Consensus code must do its own checking before invoking Ledger code that relies on the `EpochInfo` never failing.
+
+- By design, any invocation of Ledger code that causes the `EpochInfo` to return `Left` would be a _bug_, with no obvious way to recover.
+  Thus this isn't a particularly burdensome requirement; not much extra code (such as `for < maxFor` above).
+  And also pure exceptions are reasonable choice.
```

Comment:
Maybe call them "imprecise exceptions" as in the [original paper](https://www.microsoft.com/en-us/research/wp-content/uploads/1999/05/except.pdf)? Apparaently, they are sometimes also called "impure exceptions":

> A small point on terminology: in the safe-exceptions framework, imprecise exceptions are called “impure exceptions”. I am not fond of this terminology: as I’ve argued at the beginning of this post, imprecise exceptions are no more impure than infinite loops. They are just more useful.

https://www.tweag.io/blog/2020-04-16-exceptions-in-haskell/